src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/aws/AWSEC2JCloudsProvider.java

```diff
@@ -222,4 +284,69 @@ public void removeInstancePublicIp(Infrastructure infrastructure, String instanc
         });
     }
 
+    @Override
+    public SimpleImmutableEntry<String, String> createKeyPair(Infrastructure infrastructure, Instance instance) {
+        KeyPairApi keyPairApi = getKeyPairApi(infrastructure);
+        String region = getRegionFromImage(instance);
+        String keyPairName = "default-" + region + "-" + UUID.randomUUID();
+        try {
+            KeyPair keyPair = keyPairApi.createKeyPairInRegion(region, keyPairName);
+            logger.info("Created key pair '" + keyPairName + "' in region '" + region + "'");
+            return new SimpleImmutableEntry<>(keyPairName, keyPair.getKeyMaterial());
+        } catch (RuntimeException e) {
+            logger.warn("Cannot create key pair in region '" + region);
+            return null;
+        }
+    }
+
+    private KeyPairApi getKeyPairApi(Infrastructure infrastructure) {
+        ComputeService computeService = getComputeServiceFromInfastructure(infrastructure);
+        EC2Api ec2Api = computeService.getContext().unwrapApi(EC2Api.class);
+        if (ec2Api.getKeyPairApi().isPresent()) {
+            return ec2Api.getKeyPairApi().get();
+        } else {
+            throw new UnsupportedOperationException("Cannot retrieve AWS key pair API, which enables key pair creation");
+        }
+    }
+
+    private String getRegionFromImage(Instance instance) {
+        String image = instance.getImage();
+        return image.split(INSTANCE_ID_REGION_SEPARATOR)[0];
+    }
+
+    @Override
+    public RunScriptOptions getDefaultRunScriptOptions(String instanceId, Infrastructure infrastructure,
+            String instanceTag) {
+        ComputeService computeService = getComputeServiceFromInfastructure(infrastructure);
+
+        if (instanceId == null) {
+            // retrieve at least one instance that has the given tag
+            instanceId = getCreatedInfrastructureInstances(infrastructure).stream()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Introduce a new variable instead of reusing the parameter "instanceId". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1226)